Adding the final keyword for EntityManager to enforce no extension

usmanzafar opened the pull request:
Added the final keyword for EntityManager to indicate that extending EntityManager is not a valid extension point as recommended in https://github.com/doctrine/doctrine2/pull/6595 by @Ocramius / @Majkl578 
Let me know what do you guys think, very excited - first time contributing to a project that i use almost everyday for the last 3 years.

Commit:
added the final keyword EntityManager to indicate that extending EntityManager is not a valid extension point.

Majkl578 commented:
Unfortunately it won't be _that_ easy as just uncommenting the modifier. There is at least one case that needs to be adressed first: Doctrine\Tests\Mocks\EntityManagerMock currently extends EntityManager. :/

Ocramius commented:
To be added:

 * [ ] document breaking change in `UPGRADE.md` or equivalent
 * [ ] change test suite so the mock wraps a real `EntityManager` and implements the `EntityManagerInterface`
 * [ ] (maybe) consider killing `EntityManagerDecorator`?

Majkl578 committed:
Typehint EntityManagerInterface instead of EntityManager where possible

usmanzafar commented:
Yes, should have checked thats my bad. Let me have a look at this and get back asap.

Commit:
added the test and updated the entitymanager

alcaeus commented on tests/Doctrine/Tests/Mocks/EntityManagerMock.php:
```diff
@@ -23,12 +33,25 @@ class EntityManagerMock extends EntityManager
     public function getUnitOfWork()
     {
-        return isset($this->uowMock) ? $this->uowMock : parent::getUnitOfWork();
+        return isset($this->uowMock) ? $this->uowMock : $this->em->getUnitOfWork();
```
`$this->uowMock ?? $this->em->getUnitOfWork();`

alcaeus commented on tests/Doctrine/Tests/Mocks/EntityManagerMock.php:
```diff
@@ -60,7 +83,7 @@ public function setProxyFactory($proxyFactory)
     public function getProxyFactory()
     {
-        return isset($this->proxyFactoryMock) ? $this->proxyFactoryMock : parent::getProxyFactory();
+        return isset($this->proxyFactoryMock) ? $this->proxyFactoryMock : $this->em->getProxyFactory();
```
`$this->proxyFactoryMock ?? $this->em->getProxyFactory()`

usmanzafar replied:
ahh yes! PHP 7 love. Will do it now. thanks @alcaeus :)

Commit:
Merge remote-tracking branch 'Majkl578/dev/typehint-EntityManagerInterface' into add-finalkeyword-entitymanager-patch

Commit:
fixed the broken tests and updated the UPGRADE guide to notify about BC

usmanzafar commented:
Hey @alcaeus i have updated the Mock / tests as per the recommendation, i have also merged https://github.com/doctrine/doctrine2/pull/6599 by @Majkl578 as this was necessary for the tests to pass. 
Let me know your thoughts.  Thanks

Ocramius commented:
Please rebase: the interface patch was merged to `develop`

Commit:
Merge branch 'develop' into add-finalkeyword-entitymanager-patch

usmanzafar commented:
@Ocramius  - done captain!

Ocramius commented on UPGRADE.md:
```diff
@@ -6,6 +6,9 @@
 ## BC Break: Removed `EntityManager#flush($entity)` and `EntityManager#flush($entities)`
 
+Final keyword has been added to the ``EntityManager::class`` in order to ensure that EntityManager is not used as valid extension point. Valid extension point should be EntityManagerInterface.
+## BC Break: Removed `EntityManager#flush($entity)` and `EntityManager#flush($entities)`
```
Something went wrong in the rebase here, and your BC break heading is gone

Ocramius commented on tests/Doctrine/Tests/ORM/Hydration/ObjectHydratorTest.php:
```diff
@@ -1005,6 +1005,7 @@ public function testScalarQueryWithoutResultVariables($userEntityKey)
      */
     public function testCreatesProxyForLazyLoadingWithForeignKeys()
     {
+        $this->markTestSkipped('Fix with StaticProxy Factory as ProxyFactory::class does not exist anymore');
```
Skipping this is unrelated to your patch

Ocramius commented on tests/Doctrine/Tests/ORM/Hydration/ObjectHydratorTest.php:
```diff
@@ -1054,6 +1055,7 @@ public function testCreatesProxyForLazyLoadingWithForeignKeys()
      */
     public function testCreatesProxyForLazyLoadingWithForeignKeysWithAliasedProductEntity()
     {
+        $this->markTestSkipped('Fix with StaticProxy Factory as ProxyFactory::class does not exist anymore');
```
Same

Ocramius commented on tests/Doctrine/Tests/ORM/QueryBuilderTest.php:
```diff
@@ -792,7 +792,7 @@ public function testMultipleIsolatedQueryConstruction()
         $qb = $this->em->createQueryBuilder();
-        self::assertEquals($this->em, $qb->getEntityManager());
+        self::assertEquals($this->em->getEM(), $qb->getEntityManager());
     }
```
Correct the type of the `em` property in this test

Ocramius commented on tests/Doctrine/Tests/ORM/EntityManagerTest.php:
```diff
@@ -222,7 +222,7 @@ public function testTransactionalAcceptsVariousCallables()
     public function transactionalCallback($em)
     {
-        self::assertSame($this->em, $em);
+        self::assertSame($this->em->getEM(), $em);
```
Correct the type of the `em` property in this test

Commit:
change the em property to fix the tests and UPGRADE file as per the PR comments.

usmanzafar commented:
Hey @Ocramius apologies for not replying earlier, i have fixed the tests and updated others as per your advice. Let me know your thoughts on this. Thanks

Ocramius commented on lib/Doctrine/ORM/EntityManager.php:
```diff
@@ -139,7 +139,7 @@
      * @param \Doctrine\Common\EventManager $eventManager
      */
-    protected function __construct(Connection $conn, Configuration $config, EventManager $eventManager)
+    public function __construct(Connection $conn, Configuration $config, EventManager $eventManager)
```
@usmanzafar was this made `public` on purpose?

Commit:
Reverted the public method on construct to protected. Updated the tests.

usmanzafar commented:
Hey @Ocramius having that method to public was a major mistake at my end. I have updated the mock to create an EM via static method. 
Secondly, after this change to mock i was getting an error/exception of  "Cannot use different EventManager instances for EntityManager and Connection" - i have fixed that issue and updated the creation of EntityManager via Mock so that eventManager and connection's event manager remain similar where applicable. If you believe that this is outside the scope of the patch then let me know i am happy to take them away from tests and keep EntityManager / EntityManagerMock changes.
Thanks

Commit:
Merge remote-tracking branch 'upstream/develop' into add-finalkeyword-entitymanager-patch

Commit:
updated the heading for upgrade doc

Ocramius reviewed (approved):
LGTM :+1:

Ocramius commented:
> Lastly, as per the new changes merged into develop the test `OnClassMetadataNotFoundEventArgsTest::testEventArgsMutability` is creating a mock of a final class which generates a warning in phpunit - shall we ignore this or do i need to update this - let me know your thoughts ?

@usmanzafar is it related to this patch?

usmanzafar commented:
@Ocramius   Technically no, visually I was being annoyed looking at the fact that final class is being mocked  haha. I think we are good man. so happy to be part of this project. Happy to see if this gets merged to develop :)

Ocramius commented:
:ship: then!

Thanks @usmanzafar \o/
